### PR TITLE
added sse feature for rust resampler implementation

### PR DESCRIPTION
**Pull request by khodzha**
this works only on nightly though

### CODE REVIEW

**Comment by khodzha** [comment]
results for `resampling_rust` bench:
with sse: `[11.332 ms 11.391 ms 11.430 ms]`
without sse: `[17.240 ms 17.490 ms 17.915 ms]`

**Comment by khodzha** [comment]
> Alternatively the actual sse simd intrinsics could be used instead and the cpu feature gates could be used.

Could you elaborate on this please? Im not familiar with this stuff :man_shrugging:

File: resampler/src/speex.rs [hunk 1]
```diff
@@ -884,6 +936,24 @@ fn resampler_basic_interpolate_single(
     out_sample as i32
 }
 
+#[cfg(feature = "simd")]
+fn cubic_coef(frac: f32, interp: &mut f32x4) {
```
Comment (khodzha) [inline resampler/src/speex.rs:940]: maybe i should return a value from here directly?

File: resampler/src/speex.rs [hunk 2]
```diff
@@ -895,6 +965,59 @@ fn cubic_coef(frac: f32, interp: &mut [f32]) {
...
+
+    while !(last_sample >= *in_len || out_sample >= *out_len) {
+        let iptr: &[f32] = &in_0[last_sample as usize..];
+        let offset = samp_frac_num * st.oversample / st.den_rate;
+        let frac =
+            (samp_frac_num * oversample % den_rate) as f32 / den_rate as f32;
+        let mut accum = f64x4::splat(0.0);
+        iptr.iter().zip(0..n).for_each(|(&curr_in, j)| {
+            let idx = (2 + (j + 1) * oversample as usize) - offset as usize;
+            accum += f64x4::from(
```
Comment (khodzha) [inline resampler/src/speex.rs:998]: i wonder if its possible to eradicate this duplication since only few lines are changed inside these functions

**Comment by lu-zero** [comment]
> > Alternatively the actual sse simd intrinsics could be used instead and the cpu feature gates could be used.
> 
> Could you elaborate on this please? Im not familiar with this stuff 🤷‍♂️

The [core::arch](https://doc.rust-lang.org/core/arch/index.html) part of the standard library contains the arch-specific intrinsics and feature gates to check for them at runtime/compile time.

The page linked has some documentation but feel free to poke me on irc if you need more information :)

**Comment by khodzha** [comment]
rust with sse: `[9.7488 ms 9.8702 ms 10.001 ms]`
c with sse: `[7.1833 ms 7.2155 ms 7.2356 ms]`

File: resampler/src/speex.rs [hunk 3]
```diff
@@ -1358,3 +1528,86 @@ fn speex_resampler_magic<'a, 'b>(
...
+
+        assert_eq!(output, vec![1.0, 2.0, 3.0, 4.0]);
+    }
+
+    #[cfg(all(
+        any(target_arch = "x86", target_arch = "x86_64"),
+        target_feature = "sse2"
+    ))]
+    #[test]
+    #[allow(clippy::float_cmp)]
```
Comment (khodzha) [inline resampler/src/speex.rs:1602]: not sure whether this is good idea but im reluctant to introduce tolerance here :man_shrugging:

Comment (Luni-4) [reply resampler/src/speex.rs:1602]: You can do something like [this](https://github.com/rust-av/speexdsp-rs/blob/master/fft/tests/fftwrap.rs#L156) to compare floats.

Comment (khodzha) [reply resampler/src/speex.rs:1602]: yeah but i mean these are hard coded values and they dont have any error due to conversion :disappointed:

Comment (Luni-4) [reply resampler/src/speex.rs:1602]: Cast them to `usize`? I've seen they don't have the mantissa

**Comment by Luni-4** [comment]
Could you change the commit message according to the other ones and rebase, please?

**Comment by khodzha** [comment]
`[9.7546 ms 10.100 ms 10.426 ms]` on my pc

**Comment by Luni-4** [comment]
> `[9.7546 ms 10.100 ms 10.426 ms]` on my pc (i7 6700k)

I think we could try to reduce this result using iterators, in this way we can autovectorize some loops.

**Pull request merged**